File: windows/Relay.Core/QrPayload.cs
using System.Text.Json.Serialization;

namespace Relay.Core;

/// <summary>
/// Versioned pairing payload. Wire contract: /shared/qr-payload.schema.json.
/// Any change happens there first.
/// </summary>
public sealed record QrPayload
{
    public const string ModeSocks5 = "socks5";
    public const string ModeWireguard = "wireguard";

    [JsonPropertyName("v")] public required int V { get; init; }
    [JsonPropertyName("mode")] public required string Mode { get; init; }
    [JsonPropertyName("host")] public required string Host { get; init; }
    [JsonPropertyName("port")] public required int Port { get; init; }
    [JsonPropertyName("name")] public string? Name { get; init; }
    [JsonPropertyName("issuedAt")] public long? IssuedAt { get; init; }
    [JsonPropertyName("wg")] public WgParams? Wg { get; init; }
}

public sealed record WgParams
{
    [JsonPropertyName("serverPublicKey")] public required string ServerPublicKey { get; init; }
    [JsonPropertyName("clientPrivateKey")] public required string ClientPrivateKey { get; init; }
    [JsonPropertyName("allowedIps")] public required string AllowedIps { get; init; }
    [JsonPropertyName("endpointPort")] public required int EndpointPort { get; init; }
    [JsonPropertyName("dns")] public required string Dns { get; init; }
}

File: android/app/src/main/kotlin/io/relay/app/core/QrPayload.kt
package io.relay.app.core

import kotlinx.serialization.Serializable

/**
 * Versioned pairing payload. Wire contract: /shared/qr-payload.schema.json.
 * Any change here must change the schema first and bump [QrPayloadCodec.SUPPORTED_VERSION].
 */
@Serializable
data class QrPayload(
    val v: Int,
    val mode: String,
    val host: String,
    val port: Int,
    val name: String? = null,
    val issuedAt: Long? = null,
    val wg: WgParams? = null,
) {
    companion object {
        const val MODE_SOCKS5 = "socks5"
        const val MODE_WIREGUARD = "wireguard"
    }
}

@Serializable
data class WgParams(
    val serverPublicKey: String,
    val clientPrivateKey: String,
    val allowedIps: String,
    val endpointPort: Int,
    val dns: String,
)
